### PR TITLE
Enable `self.device` access in setup hook

### LINKED ISSUE #13108: `self.device` is not correct during `setup`
### To Reproduce

```python
import os

from pytorch_lightning import Trainer
from pytorch_lightning.demos.boring_classes import BoringModel


class BugReportModel(BoringModel):
    def setup(self, stage=None):
        # current workaround
        assert self.trainer.strategy.root_device.type == "cuda"
        # error! it is on cpu
        assert self.device.type == "cuda"


def run():
    model = BugReportModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        fast_dev_run=True,
        accelerator="gpu",
        devices=1,
    )
    trainer.fit(model)


if __name__ == "__main__":
    run()
```

### Expected behavior

`lightning_module.device` and `strategy.device` point to the same device

### Additional context

Reported in Slack by Joshua Chia

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #13108

This PR sets the `LightningModule.device` to the target device before running the setup hook, so that the value is available during `LightningModule.setup` even though the module parameters are not yet placed on the device.

This solution was discussed in #13108. The alternatives of changing hook order and device setup in the trainer don't seem to be an option because they would lead to breaking changes.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli

### CODE REVIEW

**Review by carmocca — APPROVED** [review]
Nice! Finally fixed

**Review by justusschock — APPROVED** [review]
